<div class="row clearfix">
    <div class="col-md-12 column">
        <div class="jumbotron">
            <h1><strong>{APPRIS}</strong></h1>
            <h2>Annotating principal splice isoforms</h2>
        </div>
    </div>
</div>
<div class="row clearfix">
    <div class="col-md-12 column">
        <tools-tpl></tools-tpl>
    </div>
</div>
<hr />
<div class="section-tout">
    <p class="lead text-center">APPRIS Database currently houses annotations for <a href="#/species">five genomes
        »</a></p>
    <div class="row row-centered">
        <div class="col-md-1"></div>
        <div class="col-md-2"
             data-ng-repeat="specie in species | showPopularGenomes | orderObjectBy:'order' ">
            <!--<a class="media-left" data-ng-href="/#/species">-->
            <a class="media-left">
                <img class="img-thumbnail" data-ng-src="img/{{specie.common | lowercase}}_logo.jpg" style="width: 65px" />
            </a>
            <div class="media-body">
                <h4 class="media-heading">{{specie.common}}</h4>
                <ul class="list-unstyled">
                    <li data-ng-repeat="assembly in specie.assemblies"><strong>Assemblies:</strong>
                        {{assembly.name}} (Ensembl{{assembly.dataset}})</em></li>
                </ul>
            </div>
        </div>
    </div>
</div>
<div class="row clearfix">
    <div class="col-md-4 column">
        <h2>APPRIS</h2>
        <p>
            The <strong>APPRIS Database</strong> [2] houses protein structure, function and conservation information
            for splice variants in five genomes. The <strong>APPRIS WebServer</strong> [1] executes a range of computational methods to annotate splice variants for individual genes. As part of its annotation process APPRIS selects a single CDS as the principal isoform for each gene.
        </p>
        <p>
            <a href="#/help/database">View details »</a>
        </p>
    </div>
    <div class="col-md-4 column">
        <h2>Publications</h2>
        <address>
            [1] <strong>APPRIS WebServer and WebServices.</strong>
            <abbr title="Rodriguez JM, Carro A, Valencia A, and Tress ML">Rodriguez JM, et al.</abbr><br />
            Nucleic Acids Res. 2015 May 18.<br/>
            PUBMED:<a href="http://ukpmc.ac.uk/abstract/MED/25990727" target="_blank">25990727</a> DOI:<a
                href="http://dx.doi.org/10.1093/nar/gkv512" target="_blank">10.1093/nar/gkv512</a>
        </address>
        <address>
            [2] <strong>APPRIS: annotation of principal and alternative splice isoforms.</strong>
            <abbr title="Rodriguez JM, Maietta P, Ezkurdia I, Pietrelli A, Wesselink JJ, Lopez G, Valencia A and Tress ML">Rodriguez JM, et al.</abbr><br />
            Nucleic acids research 2013;41;Database issue;D110-7<br/>
            PUBMED:<a href="http://ukpmc.ac.uk/abstract/MED/23161672" target="_blank">23161672</a> PMC:<a href="http://ukpmc.ac.uk/articles/PMC3531113" target="_blank">3531113</a> DOI:<a href="http://dx.doi.org/10.1093/nar/gks1058">10.1093/nar/gks1058</a>
        </address>
        <p>
            <a href="#/publications">View details »</a>
        </p>
    </div>
    <div class="col-md-4 column">
        <h2>API</h2>
        <p>
            The APPRIS Database and APPRIS WebServer can be accessed through a range of RESTful web services. These services can retrieve both gene/transcript information and the results of the individual APPRIS modules.
        </p>
        <p>
            <a href="#/help/services">View details »</a>
        </p>
    </div>
</div>