<div>
    <!--<help-methods></help-methods>-->
    <p>The “APPRIS annotations” section shows the scores of the <a href="#/help/database">APPRIS modules</a>.</p>
    <p>
        <img src="img/report-annots.png" class="img-thumbnail img-responsive img-center" style="width: 80%">
    </p>
    <h4>Num. Functional Residues</h4>
    <p>firestar predicts functionally important residues. The box shows the total number of functional residues detected by firestar for each transcript.</p>

    <h4>Tertiray Structure Score</h4>
    <p>Matador3D analyses protein structural information for each variant. The score is based on the number of exons that can be mapped to structural homologues. This score is only comparable for transcripts within the same gene and should not be used to compare between genes.</p>

    <h4>Conservation score (vertebrates)</h4>
    <p>CORSAIR carries out BLAST searches against vertebrates to determine the most likely principal isoform. The score approximates to the number of species that have isoforms that can be aligned to each tested isoform without introducing gaps.</p>

    <h4>Domains score</h4>
    <p>SPADE identifies the functional domains present in a transcript and detects those domains that are damaged
        (not whole). The number represents the sum of bitscores in Pfam alignment.</p>

    <h4>Num. Transmembrane Helices</h4>
    <p>THUMP makes unanimous predictions of trans-membrane helices. The score shows the number of trans-membrane helices predicted by THUMP for each variant.</p>

    <h4>Signal Peptide Sequences and Mitochondrial Signal Sequences</h4>
    <p>CRASH predicts the presence and location of signal peptides and cleavage sites in amino acid sequences using an especially reliable form of SignalP and TargetP. Isoforms that are predicted to have reliable signal sequences are marked.</p>

</div>
