<div>
    <p>Alternative splicing generates different gene products. Recent studies have estimated that almost 100% of multi-exon human genes produce differently spliced mRNAs. It is important to designate one of the isoforms as the principal isoform in order to predict the potential changes in function, structure or localisation brought about by alternative splicing.
    </p>
    <p>APPRIS annotates variants with biological data such as <strong>protein structural information</strong>, <strong>functionally important residues</strong>, <strong>conservation of functional domains</strong>and <strong>evidence of cross-species conservation</strong>. The APPRIS Database selects
        a <strong>principal isoforms</strong> based on this evidence.
    </p>
    <div class="thumbnail">
        <img src="img/APPRIS_analysis.png" class="img-thumbnail img-responsive img-center" style="width: 90%">
        <div class="caption">
            <h5>APPRIS Database analysis</h5>
            <small>The APPRIS Database automates a range of computational methods that are used to annotate alternative splice variants and to define principal variants. The splice isoform annotations are the results of the six modules in the APPRIS Database; the final module selects the principal isoforms.</small>
        </div>
    </div>

    <h3>Method Description</h3>
    <help-methods></help-methods>

    <h2>Report view, Visualising the annotations</h2>
    <help-report></help-report>
    <hr>

</div>
